sdk/eventhub/Azure.Messaging.EventHubs/src/Core/TransportProducerPool.cs

```diff
@@ -0,0 +1,396 @@
+                {
+                    RemoveAfter = removeAfter.Value;
+                }
+            }
+        }
+
+        /// <summary>
+        ///   A class wrapping a <see cref="Core.TransportProducerPool"/>, triggering a clean-up when the object is disposed.
+        /// </summary>
+        ///
+        internal class PooledProducer: IAsyncDisposable
+        {
+            /// <summary>
+            ///   A function responsible of cleaning up the resources in use.
+            /// </summary>
+            ///
+            private Func<Task> CleanUp { get; }
+
+            /// <summary>
+            ///   The <see cref="Core.TransportProducerPool"/> that is associated with the current <see cref="PooledProducer"/>.
+            /// </summary>
+            ///
+            public TransportProducerPool TransportProducerPool { get; }
+
+            /// <summary>
+            ///   The unique identifier of a partition associated with the Event Hub.
+            /// </summary>
+            ///
+            public string PartitionId { get; set; }
+
+            /// <summary>
+            ///   Initializes a new instance of the <see cref="PooledProducer"/> class.
+            /// </summary>
+            ///
+            /// <param name="transportProducerPool">An Event Hub transport-specific producer specific to a given partition.</param>
+            /// <param name="partitionId">The unique identifier of a partition associated with the Event Hub.</param>
+            /// <param name="cleanUp">The function responsible of cleaning up the resources in use.</param>
+            ///
+            public PooledProducer(TransportProducerPool transportProducerPool,
+                                  string partitionId = default,
+                                  Func<Task> cleanUp = default)
+            {
+                Argument.AssertNotNull(transportProducerPool, nameof(transportProducerPool));
+
+                TransportProducerPool = transportProducerPool;
+                CleanUp = cleanUp;
+                PartitionId = partitionId;
+            }
+
+            /// <summary>
+            ///   Creates a <see cref="TransportProducer"/> matching the partition id passed as input.
+            /// </summary>
+            ///
+            /// <param name="connection">The <see cref="EventHubConnection" /> connection to use for communication with the Event Hubs service.</param>
+            /// <param name="retryPolicy">The policy to use for determining retry behavior for when an operation fails.</param>
+            ///
+            /// <returns>A <see cref="TransportProducer"/> matching the partition id passed as input.</returns>
+            ///
+            public TransportProducer GetTransportProducer(EventHubConnection connection,
+                                                          EventHubsRetryPolicy retryPolicy)
+            {
+                return TransportProducerPool.GetTransportProducer(PartitionId, connection, retryPolicy);
```

Comment:
@jsquire @kinelski @ShivangiReja 

I did have a ```PooledProducer``` wrapping a ```TransportProducerPool```.

The impediment I faced is that we cannot add ```null``` keys to the ```ConcurrentDictionary<TKey, TValue>``` so some logic was needed to return the right producer.

This logic is in the ```TransportProducerPool.GetTransportProducer``` method.

Reply:
I may have gotten lost in the history of details, so I'm going to take a step or three back and let you call me out where I'm being silly.   Here's what I'd think of as the user experience for the pool:

```csharp
// Somewhere in my constructor.
TransportProducerPool pool = new TransportProducerPool(connection, retryPolicy);

// When I want to send a thing.
await using (var pooledProducer = pool.GetProducer(partitionId))
{
    try
    {
        await  pooledProducer.TransportProducer.SendAsync(myBatch).ConfigureAwait(false);
    }
    catch (  RETRY STUFF HERE for "the producer is closed," )
    {
    }
}
```

There as a bunch of book keeping and another level of wrapping that I think we needed to track things and reduce the potential for a race condition, but we seem to have a lot of `public` members for creating the transport producer at different levels.    I question if we're exposing our implementation details.

Can you remind me of what I'm missing?   I'm sure that we talked over a bunch of this and I'm just coming at this after being focused elsewhere and got lost a bit in the details.

Reply:
> Can you remind me of what I'm missing? I'm sure that we talked over a bunch of this and I'm just coming at this after being focused elsewhere and got lost a bit in the details.

You are a genius. That definitely works.

Refactored.

Reply:
You are far too kind; most days, I'm a barely functioning adult.   😄